src/ray/gcs/gcs_client/service_based_accessor.h

```diff
@@ -9,6 +10,41 @@ namespace gcs {
 
 class ServiceBasedGcsClient;
 
+/// \class Sequencer
+/// Sequencer guarantees that all operations with the same ordering key are sequenced.
+/// This class is thread safe.
+template <class KEY>
+class Sequencer {
+ public:
+  void execute_ordered(KEY key, std::function<void()> operation) {
+    mutex_.lock();
+    pending_operations_.insert(value_type(key, operation));
+    pending_operations_[key].push_back(operation);
+    int queue_size = pending_operations_[key].size();
+    mutex_.unlock();
+    if (1 == queue_size) {
+      operation();
+    }
+  }
+
+  void post_execute(KEY key) {
+    mutex_.lock();
+    pending_operations_[key].pop_front();
+    if (pending_operations_[key].empty()) {
+      pending_operations_.erase(key);
+      mutex_.unlock();
+    } else {
+      auto operation = pending_operations_[key].front();
+      mutex_.unlock();
+      operation();
+    }
+  }
+
+ private:
+  std::mutex mutex_;
```

Comment:
use `absl::Mutex` and `GUARDED_BY` for better readability. There are a few such examples in the code base, you can search these keywords.

Reply:
fixed